### PR TITLE
ydoc-server is a separate module

### LINKED ISSUE #10155: ydoc should be a separate JPMS module
Currently, `org.enso.syntax2.Parser` is present both in `runtime.jar` and in `runner.jar`. This causes issues with the class being loaded by two class loaders. Let's remove that class from these fat jars, and provide it as a separate module on module-path. Also consider providing a separate `ydoc-server` module (currently, ydoc-server classes are inside runner.jar).

Priority is high, as it blocks cloud's integration of ydoc-server

```[tasklist]
### Tasks
- [ ] ydoc-server is separate module
``` 

Related:
- https://github.com/enso-org/enso/pull/10123

### PR DESCRIPTION
**Pull request by Akirathan**
Fixes #10155

### Important Notes

- Remove remnants of deprecated Scala parser
- The following projects are now JPMS modules provided on system module-path (in `components` directory):
  - `ydoc-server`
  - `profiling-utils`
  - `syntax-rust-definition`
- The contents of the aforementioned modules are excluded from both `runner.jar` and `runtime.jar` fat jars.
- Suggestions are serialized and deserialized with our `Persistance` framework, rather than via the default Java `OutputObjectWriter`.
  - See [this comment](https://github.com/enso-org/enso/pull/10156#issuecomment-2147703054) explaining the motivation for serialization changes.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-fat-jar/src/main/java/module-info.java [hunk 1]
```diff
@@ -3,6 +3,11 @@
   requires java.net.http;
   // Because of akka.util.Unsafe
   requires jdk.unsupported;
+  requires org.enso.syntax;
+  // profiling-utils is not needed for runtime. It is needed for runner.jar,
+  // but we have to include it here so that the forwarding in IsolatedClassLoader
+  // works.
+  requires org.enso.profiling;
```
Comment (Akirathan) [inline engine/runtime-fat-jar/src/main/java/module-info.java:10]: Removal of these unnecessary dependencies tracked in https://github.com/enso-org/enso/issues/10157

File: engine/runtime/src/main/java/org/enso/ClassLoaderConstants.java [hunk 2]
```diff
@@ -13,10 +13,7 @@ public class ClassLoaderConstants {
    * consistent.
    */
   public static final List<String> CLASS_DELEGATION_PATTERNS =
-      List.of("org.graalvm", "java", "org.slf4j", "ch.qos", "io.helidon");
-
-  public static final List<String> RESOURCE_DELEGATION_PATTERNS =
-      List.of("org.slf4j", "ch.qos", "io.helidon");
+      List.of("org.graalvm", "java", "org.slf4j", "ch.qos", "io.helidon", "org.enso.profiling", "org.enso.syntax2", "org.enso.ydoc");
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/ClassLoaderConstants.java]: Only one field is necessary here. From Classloader's perspective, a class is a special case of a resource when looking up via a binary name.

**Comment by Akirathan** [comment]
On https://github.com/enso-org/enso/pull/10156/commits/9de7c282433a7a1de6502bfaefe4547a6717d06b, ydoc can be started and it seems to synchronize correctly. But SuggestionsCache fail to load with "filter status REJECTED":
```
[INFO] [2024-06-04T13:22:59.057] [enso.org.enso.interpreter.caches.Cache] Deserializing suggestions for library Standard.Base
[INFO] [2024-06-04T13:22:59.057] [enso.org.enso.interpreter.caches.Cache] helidon.serialFilter.pattern: javax.management.**;java.lang.**;java.rmi.**;javax.security.auth.Subject;!*
[WARN] [2024-06-04T13:22:59.059] [enso.org.enso.interpreter.caches.Cache] `Standard.Base` in /home/pavel/dev/enso/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Base/0.0.0-dev/.enso/cache/suggestions/0.0.0-dev/Standard/Base.suggestions failed to load: 
java.io.InvalidClassException: filter status: REJECTED
	at java.base/java.io.ObjectInputStream.filterCheck(ObjectInputStream.java:1439)
	at java.base/java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:2071)
	at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1927)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2252)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1762)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:540)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:498)
	at org.enso.runtime/org.enso.interpreter.caches.SuggestionsCache.deserialize(SuggestionsCache.java:81)
	at org.enso.runtime/org.enso.interpreter.caches.SuggestionsCache.deserialize(SuggestionsCache.java:27)
	at org.enso.runtime/org.enso.interpreter.caches.Cache.loadCacheFrom(Cache.java:277)
	at org.enso.runtime/org.enso.interpreter.caches.Cache.lambda$load$0(Cache.java:209)
	at java.base/java.util.Optional.flatMap(Optional.java:289)
	at org.enso.runtime/org.enso.interpreter.caches.Cache.load(Cache.java:192)
	at org.enso.runtime/org.enso.interpreter.runtime.TruffleCompilerContext.loadCache(TruffleCompilerContext.java:201)
	at org.enso.runtime/org.enso.interpreter.runtime.TruffleCompilerContext.deserializeSuggestionsImpl(TruffleCompilerContext.java:618)
	at org.enso.runtime/org.enso.interpreter.runtime.TruffleCompilerContext.deserializeSuggestions(TruffleCompilerContext.java:605)
	at org.enso.runtime/org.enso.interpreter.instrument.job.DeserializeLibrarySuggestionsJob.run(DeserializeLibrarySuggestionsJob.scala:38)
	at org.enso.runtime/org.enso.interpreter.instrument.job.DeserializeLibrarySuggestionsJob.run(DeserializeLibrarySuggestionsJob.scala:16)
	at org.enso.runtime/org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:135)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:106)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:140)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:131)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:745)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:519)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:500)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:102)
[INFO] [2024-06-04T13:22:59.060] [enso.org.enso.interpreter.caches.Cache] Invalidated the cache at [/home/pavel/dev/enso/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Base/0.0.0-dev/.enso/cache/suggestions/0.0.0-dev/Standard/Base.suggestions.meta].
[INFO] [2024-06-04T13:22:59.061] [enso.org.enso.interpreter.caches.Cache] Invalidated the cache at [/home/pavel/dev/enso/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Base/0.0.0-dev/.enso/cache/suggestions/0.0.0-dev/Standard/Base.suggestions].
[INFO] [2024-06-04T13:22:59.061] [enso.org.enso.interpreter.caches.Cache] Unable to load a cache [Standard.Base]
[WARN] [2024-06-04T13:22:59.061] [enso.org.enso.interpreter.runtime.SerializationPool] Unable to load suggestions for library [Standard.Base].
[DEBUG] [2024-06-04T13:22:59.061] [enso.org.enso.interpreter.service.ExecutionService] Job DeserializeLibrarySuggestionsJob(Standard.Base) finished in 42 ms.
```

Apparently, this is caused by helidon disabling reflection and https://github.com/enso-org/enso/pull/9528/commits/f508ffac632a7568a7804e067f9c0a8e7e277b97#r1587057920 should resolve the issue. But I have tried to set this `helidon.serialFilter.pattern` system property (as can be seen in the log) and the issue is still there. The diff is:
```diff
diff --git a/lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java b/lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java
index d5e7b1ad1..9d8dc5cb9 100644
--- a/lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java
+++ b/lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java
@@ -77,6 +77,12 @@ public final class Ydoc implements AutoCloseable {
     }
 
     public Ydoc build() {
+      // helidon disables reflection. The following property is a workaround for that, so that
+      // we can use our deserialization.
+      System.setProperty(
+          "helidon.serialFilter.pattern",
+          "javax.management.**;java.lang.**;java.rmi.**;javax.security.auth.Subject;!*");
+
       if (executor == null) {
         executor =
             Executors.newSingleThreadScheduledExecutor(
```

Let's try to fix this by serializing and deserializing suggestions with Persistance.

**Comment by Akirathan** [comment]
As of https://github.com/enso-org/enso/pull/10156/commits/2f29a5f9d1ab5cee0fc32b530b368bbbbfc8a49f, the suggestions are correctly serialized and deserialized using our `Persistance` framework.  I have just tried that on opening a dummy project. Also, the resulting `*.suggestions` binary file is slightly smaller than it used to be, as now, we don't serialize all the sources of a library anymore.

Let's continue with simple tests of ydoc server...

File: engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java [hunk 3]
```diff
@@ -86,37 +72,6 @@ public void tearDown() {
     ctx.close();
   }
 
-  @Test
-  public void testMostRuntimeExceptionsCanPropagateFromBuiltinMethods() {
```
Comment (4e6) [inline engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java:90]: Why do we want to change that? Seems irrelevant

Comment (Akirathan) [reply engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java:90]: See commit https://github.com/enso-org/enso/pull/10156/commits/1809ec0baae702496c4dafa3740b29bb821ecb09. After changing module structure, I was unable to make this test working in `interpreter-dsl-test` no matter what I tried. At the end, I have decided that I have no idea what this test should test anyway, so I have just decided to remove it. Note that there is still a relevant test in this file.

File: build.sbt [hunk 4]
```diff
@@ -610,26 +615,6 @@ lazy val compileModuleInfo = taskKey[Unit]("Compiles `module-info.java`")
 // === Internal Libraries =====================================================
 // ============================================================================
 
-lazy val `syntax-definition` =
-  project in file("lib/scala/syntax/definition")
```
Comment (JaroslavTulach) [inline build.sbt:614]: Cleanup is good.

File: build.sbt [hunk 5]
```diff
@@ -596,10 +595,16 @@ lazy val componentModulesPaths =
...
   val thirdPartyModFiles = thirdPartyMods.map(_.data)
   val arrow              = (`runtime-language-arrow` / Compile / packageBin).value
+  val syntax             = (`syntax-rust-definition` / Compile / packageBin).value
+  val ydoc               = (`ydoc-server` / Compile / packageBin).value
+  val profilingUtils     = (`profiling-utils` / Compile / packageBin).value
   val runtime            = (`runtime-fat-jar` / assembly / assemblyOutputPath).value
   val ourMods = Seq(
     runtime,
-    arrow
+    arrow,
```
Comment (JaroslavTulach) [inline build.sbt:604]: Shouldn't we compute `outMods` by scanning projects that contain `javaModuleName`?

File: docs/sbt-cheatsheet.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
+# sbt cheatsheet
+
+This document contains various references and notes about sbt build definition.
+One of the motivation for this document is that the
+[official sbt documentation](https://www.scala-sbt.org/1.x/docs/index.html) is
+not very informative.
```
Comment (JaroslavTulach) [inline docs/sbt-cheatsheet.md:6]: Good idea.

File: engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java (hunk 3 above)
Comment (JaroslavTulach) [inline engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java:90]: How this is related to `ydoc-server`?

File: engine/runtime/src/main/java/org/enso/interpreter/caches/SuggestionsCache.java [hunk 7]
```diff
@@ -146,56 +136,13 @@ public String blobHash(Metadata meta) {
...
-
-    public List<Suggestion> getSuggestions() {
-      return suggestions.getSuggestions();
-    }
-  }
+  /**
+   * @param libraryName
+   * @param suggestions Must not be null.
+   */
+  public record CachedSuggestions(LibraryName libraryName, ArrayList<Suggestion> suggestions) {}
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/caches/SuggestionsCache.java:143]: Shouldn't we prefer `java.util.List` over the implementation type?

File: lib/scala/syntax/README.md [hunk 8]
```diff
@@ -1,3 +0,0 @@
-# Syntax
-
-These two libraries define and implement the parser for Enso's syntax.
```
Comment (JaroslavTulach) [inline lib/scala/syntax/README.md:3]: Good cleanup.

**Review by JaroslavTulach — APPROVED** [review]
Possibly we could test that `libenso_parser` isn't in any of the JARs. And we could test that there are no duplicit classes among `runtime.jar` and `runner.jar` (except finite list). Otherwise the changes look sane.

File: engine/runtime/src/main/java/org/enso/interpreter/caches/SuggestionsCache.java (hunk 7 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/caches/SuggestionsCache.java:143]: I have tried that with `java.util.List`, but it does not work, since it is an interface, and `persistance-dsl` does not accept interfaces - it fails with an error message like: "java.util.List must contain one public parameterless constructor".

File: engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java (hunk 3 above)
Comment (Akirathan) [reply engine/interpreter-dsl-test/src/test/java/org/enso/interpreter/dsl/test/ThrowableCatchTest.java:90]: See https://github.com/enso-org/enso/pull/10156#discussion_r1630975940. `interpreter-dsl-test` got broken because of the changes with modules. Moreover, `interpreter-dsl-test` project was never setup to properly work with modules, the tests there were invoked with plain old class-path.

File: build.sbt (hunk 5 above)
Comment (Akirathan) [reply build.sbt:604]: This is not a bad idea. But it won't save us a lot of manual work, since we will still need to explicitly get modules inside module-path in various other places, like `runtime-integration-tests`, `language-server/test` or `project-manager/test`, and others.
